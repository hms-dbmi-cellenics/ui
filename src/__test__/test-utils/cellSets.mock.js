const mockCellSets = {
  initialLoadPending: false,
  error: false,
  loading: false,
  accessible: true,
  hierarchy: [
    {
      key: 'louvain',
      children: [{ key: 'cluster-a' }, { key: 'cluster-b' }, { key: 'cluster-c' }],
      cellIds: new Set(),
    },
    {
      key: 'sample',
      children: [{ key: 'sample-1' }, { key: 'sample-2' }],
      cellIds: new Set(),
    },
    {
      key: 'scratchpad',
      children: [{ key: 'scratchpad-a' }],
      cellIds: new Set(),
    },
  ],
  properties: {
    'cluster-a': {
      name: 'cluster a',
      key: 'cluster-a',
      cellIds: new Set([0, 1, 6]),
      color: '#01FFFF',
    },
    'cluster-b': {
      name: 'cluster b',
      key: 'cluster-b',
      cellIds: new Set([2, 3]),
      color: '#23FFFF',
    },
    'cluster-c': {
      name: 'cluster c',
      key: 'cluster-c',
      cellIds: new Set([4, 5]),
      color: '#45FFFF',
    },
    'sample-1': {
      name: 'Sample 1',
      key: 'sample-1',
      cellIds: new Set([0, 1, 2]),
      color: '#012FFF',
    },
    'sample-2': {
      name: 'Sample 2',
      key: 'sample-2',
      cellIds: new Set([3, 4, 5]),
      color: '#345FFF',
    },
    'scratchpad-a': {
      cellIds: new Set(['5']),
      key: 'scratchpad-a',
      name: 'New Cluster',
      color: '#5FFFFF',
    },
    louvain: {
      cellIds: new Set(),
      name: 'Louvain clusters',
      key: 'louvain',
      type: 'cellSets',
      rootNode: true,
    },
    sample: {
      cellIds: new Set([6, 7, 8]),
      name: 'Samples',
      key: 'sample',
      type: 'metadataCategorical',
      rootNode: true,
    },
    scratchpad: {
      cellIds: new Set(),
      name: 'Scratchpad',
      key: 'scratchpad',
      type: 'cellSets',
      rootNode: true,
    },
  },
};

export {
  // eslint-disable-next-line import/prefer-default-export
  mockCellSets,
};
