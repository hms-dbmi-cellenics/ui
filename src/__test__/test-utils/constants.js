export default {
  EXPERIMENT_ID: 'testae48e318dab9a1bd0bexperiment',
  CLONED_EXPERIMENT_ID: 'testae48e318dab9a1bd0bexperclone',
  PROJECT_WITH_SAMPLES_UUID: '380c53b4-mock-test-8091-4c82336d6d49',
  PROJECT_WITHOUT_SAMPLE_UUID: '4761594b-mock-test-ba89-c2c812d39bb5',
  EXPERIMENT_NAME: 'Test Experiment',
  PROJECT_ID: 'test24b6-8600-test-mock-63822d2fmock',
  SAMPLE_ID: 'test9188-d682-test-mock-cb6d644cmock',
  API_ENDPOINT: /^http?:\/\/localhost:3000.*$/,
  S3_ENDPOINT: 'http://mock.s3.amazonaws.com',
  // pragma: allowlist secret
  JWT_TOKEN: 'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6Ikpfake.eyJzdWIiOiIxMjM0NTY3ODkwIiwifakeZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTfakeM5MDIyfQ.fakexwRfakeKKF2QT4fwpMeJf36POk6yJV_adQsfake',
  MOCK_DATETIME: '0000-00-00T00:00:00.000Z',
  ETAG: '380c53b4-mock-etag-8091-4c82336d6d49',
};
