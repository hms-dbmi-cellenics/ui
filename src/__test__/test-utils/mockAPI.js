import _ from 'lodash';

// Default platform data
import cellSetsData from '__test__/data/cell_sets.json';
import backendStatusData from '__test__/data/backend_status.json';
import processingConfigData from '__test__/data/processing_config.json';
import mockDemoExperiments from '__test__/test-utils/mockData/mockDemoExperiments.json';

import fake from '__test__/test-utils/constants';

import {
  responseData,
} from '__test__/test-utils/mockData';

const promiseResponse = (
  response,
  options = {},
) => Promise.resolve(new Response(response, options));

const statusResponse = (code, body) => (
  Promise.resolve({
    status: code,
    body: JSON.stringify(body),
  })
);

const delayedResponse = (response, delay = 10000, options = {}) => new Promise((resolve) => {
  setTimeout(() => {
    resolve(new Response(response, options));
  }, delay);
});

const workerDataResult = (data) => Promise.resolve(_.cloneDeep(data));

const fetchWorkMock = (
  mockedResults,
) => ((experimentId, body) => {
  if (typeof mockedResults[body.name] === 'function') {
    return mockedResults[body.name]();
  }

  return workerDataResult(mockedResults[experimentId]);
});

const generateDefaultMockAPIResponses = (experimentId) => ({
  [`experiments/${experimentId}$`]: () => promiseResponse(
    JSON.stringify(responseData.experiments.find(({ id }) => id === experimentId)),
  ),
  [`experiments/${experimentId}/processingConfig$`]: () => promiseResponse(
    JSON.stringify(processingConfigData),
  ),
  [`experiments/${experimentId}/cellSets$`]: () => promiseResponse(
    JSON.stringify(cellSetsData),
  ),
  [`experiments/${experimentId}/backendStatus$`]: () => promiseResponse(
    JSON.stringify(backendStatusData),
  ),
  experiments$: () => promiseResponse(
    JSON.stringify(responseData.experiments),
  ),
  [`experiments/${experimentId}/samples$`]: () => promiseResponse(
    JSON.stringify(responseData.samples[0]),
  ),
  '/v2/experiments/examples$': () => promiseResponse(
    JSON.stringify(mockDemoExperiments),
  ),
  'experiments/clone$': () => promiseResponse(
    JSON.stringify(fake.CLONED_EXPERIMENT_ID),
  ),
  [`/v2/workRequest/${experimentId}`]: () => statusResponse(200, 'OK'),
});

const mockAPI = (apiMapping) => (req) => {
  const path = req.url;

  const key = _.find(Object.keys(apiMapping), (matcherStr) => new RegExp(matcherStr).test(path));

  if (!key) {
    return statusResponse({
      status: 404,
      body: `Path ${path} is undefined`,
    });
  }

  return apiMapping[key](req);
};

export default mockAPI;
export {
  generateDefaultMockAPIResponses,
  promiseResponse,
  statusResponse,
  delayedResponse,
  workerDataResult,
  fetchWorkMock,
};
