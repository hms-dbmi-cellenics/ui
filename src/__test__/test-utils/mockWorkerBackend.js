import * as socketConnectionMocks from 'utils/socketConnection';
import * as fetchWorkResponseMocks from 'utils/work/fetchWork';

import SocketMock from 'socket.io-mock';

const socketMock = new SocketMock();

jest.mock('utils/work/fetchWork', () => {
  const mockSeekFromS3 = jest.fn();
  const originalModule = jest.requireActual('utils/work/fetchWork');

  return {
    __esModule: true, // Use it when dealing with esModules
    ...originalModule,
    fetchWork: mockSeekFromS3,
  };
});

jest.mock('aws-amplify', () => ({
  configure: () => ({
    Storage: {
      AWSS3: {
        bucket: 'biomage-originals-test',
      },
    },
  }),
  Storage: {
    get: async (ETag) => `http://mock.s3.amazonaws.com/${ETag}`,
  },
  Auth: {
    federatedSignIn: jest.fn(),
    signOut: jest.fn(),
    currentAuthenticatedUser: jest.fn().mockImplementation(async () => ({
      username: 'mockuser',
      attributes: {
        email: 'mock@user.name',
        name: 'Mocked User',
      },
    })),
    currentSession: () => ({
      getIdToken: (() => ({
        getJwtToken: () => 'fakeJwtToken',
      })),
    }),
  },
  Hub: {
    listen: jest.fn(),
  },
}));

jest.mock('utils/socketConnection', () => {
  const mockEmit = jest.fn();
  const mockOn = jest.fn();
  return {
    __esModule: true,
    default: new Promise((resolve) => {
      resolve({ emit: mockEmit, on: mockOn, id: '1234' });
    }),
    mockEmit,
    mockOn,
  };
});

// To mock worker response, modify the response returned from the API
// Set to null to force fetch from API
fetchWorkResponseMocks.fetchWork.mockImplementation(() => null);

// Set up socket emitter mock
socketConnectionMocks.mockEmit.mockImplementation((workRequestType, requestBody) => {
  const responseBody = { response: { error: false } };

  // After emitting, send reply to listener
  socketMock.socketClient.emit(`WorkResponse-${requestBody.ETag}`, responseBody);
});

// Set up socket listener mock
socketConnectionMocks.mockOn.mockImplementation((channel, f) => {
  socketMock.on(channel, (responseBody) => f(responseBody));
});
