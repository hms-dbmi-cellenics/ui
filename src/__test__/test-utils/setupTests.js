import '@testing-library/jest-dom/extend-expect';
import preloadAll from 'jest-next-dynamic';

import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('localforage');
jest.mock('utils/pushNotificationMessage');

// This is needed, because the unit tests don't register the Vitessce imports
jest.mock('components/data-exploration/DynamicVitessceWrappers', () => ({
  Scatterplot: () => 'Mocked Scatterplot',
  Heatmap: () => 'Mocked Heatmap',
  Spatial: () => 'Mocked Spatial',
}));

beforeAll(async () => {
  // Add stuff that needs to run once, before all tests
  await preloadAll();
});

beforeEach(async () => {
  // Add stuff that needs to run before each test
});
