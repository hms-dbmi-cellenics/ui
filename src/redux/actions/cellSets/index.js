import createCellSet from './createCellSet';
import deleteCellSet from './deleteCellSet';
import deleteCellClass from './deleteCellClass';
import runCellSetsClustering from './runCellSetsClustering';
import runCellSetsAnnotation from './runCellSetsAnnotation';
import runCellCycleScoring from './runCellCycleScoring';
import updateCellSetsClustering from './updateCellSetsClustering';

import loadCellSets from './loadCellSets';
import reorderCellSet from './reorderCellSet';
import updateCellSetProperty from './updateCellSetProperty';
import updateCellSetSelected from './updateCellSetSelected';
import setCellSetHiddenStatus from './setCellSetHiddenStatus';
import unhideAllCellSets from './unhideAllCellSets';

export {
  createCellSet,
  deleteCellSet,
  deleteCellClass,
  runCellSetsClustering,
  runCellSetsAnnotation,
  updateCellSetsClustering,
  loadCellSets,
  reorderCellSet,
  updateCellSetSelected,
  updateCellSetProperty,
  setCellSetHiddenStatus,
  unhideAllCellSets,
  runCellCycleScoring,
};
