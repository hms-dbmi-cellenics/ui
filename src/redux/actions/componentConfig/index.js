import loadPlotConfig from 'redux/actions/componentConfig/loadPlotConfig';
import loadComponentConfig from 'redux/actions/componentConfig/loadComponentConfig';
import updatePlotConfig from 'redux/actions/componentConfig/updatePlotConfig';
import savePlotConfig from 'redux/actions/componentConfig/savePlotConfig';
import fetchPlotDataWork from 'redux/actions/componentConfig/fetchPlotDataWork';
import updatePlotData from 'redux/actions/componentConfig/updatePlotData';
import resetPlotConfig from 'redux/actions/componentConfig/resetPlotConfig';

export {
  loadPlotConfig,
  loadComponentConfig,
  updatePlotConfig,
  savePlotConfig,
  fetchPlotDataWork,
  updatePlotData,
  resetPlotConfig,
};
