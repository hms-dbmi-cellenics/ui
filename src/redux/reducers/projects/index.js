import {
  PROJECTS_CREATE,
  PROJECTS_UPDATE,
  PROJECTS_SET_ACTIVE,
  PROJECTS_DELETE,
  PROJECTS_SAVING,
  PROJECTS_SAVED,
  PROJECTS_ERROR,
  PROJECTS_METADATA_CREATE,
  PROJECTS_METADATA_UPDATE,
  PROJECTS_METADATA_DELETE,
  PROJECTS_LOADED,
  PROJECTS_LOADING,
} from '../../actionTypes/projects';

import initialState from './initialState';
import projectsCreate from './projectsCreate';
import projectsUpdate from './projectsUpdate';
import projectsDelete from './projectsDelete';
import projectsSetActive from './projectsSetActive';
import projectsSaving from './projectsSaving';
import projectsSaved from './projectsSaved';
import projectsError from './projectsError';
import projectsMetadataCreate from './projectsMetadataCreate';
import projectsMetadataUpdate from './projectsMetadataUpdate';
import projectsMetadataDelete from './projectsMetadataDelete';
import projectsLoaded from './projectsLoaded';
import projectsLoading from './projectsLoading';

import {
  SAMPLES_CREATE, SAMPLES_DELETE_API_V2,
} from '../../actionTypes/samples';

import samplesCreate from './samplesCreate';
import samplesDelete from './samplesDelete';

const projectsReducer = (state = initialState, action) => {
  switch (action.type) {
    case PROJECTS_CREATE: {
      return projectsCreate(state, action);
    }

    case PROJECTS_UPDATE: {
      return projectsUpdate(state, action);
    }

    case PROJECTS_SET_ACTIVE: {
      return projectsSetActive(state, action);
    }

    case PROJECTS_DELETE: {
      return projectsDelete(state, action);
    }

    case PROJECTS_SAVING: {
      return projectsSaving(state, action);
    }

    case PROJECTS_SAVED: {
      return projectsSaved(state);
    }

    case PROJECTS_ERROR: {
      return projectsError(state, action);
    }

    case PROJECTS_METADATA_CREATE: {
      return projectsMetadataCreate(state, action);
    }

    case PROJECTS_METADATA_UPDATE: {
      return projectsMetadataUpdate(state, action);
    }

    case PROJECTS_METADATA_DELETE: {
      return projectsMetadataDelete(state, action);
    }

    case PROJECTS_LOADED: {
      return projectsLoaded(state, action);
    }

    case PROJECTS_LOADING: {
      return projectsLoading(state);
    }

    case SAMPLES_CREATE: {
      return samplesCreate(state, action);
    }

    case SAMPLES_DELETE_API_V2: {
      return samplesDelete(state, action);
    }

    default: {
      return state;
    }
  }
};

export default projectsReducer;
