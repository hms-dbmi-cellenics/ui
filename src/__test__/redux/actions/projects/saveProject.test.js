import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';
import pushNotificationMessage from 'utils/pushNotificationMessage';
import initialProjectState from 'redux/reducers/projects/initialState';
import { saveProject } from 'redux/actions/projects';
import {
  PROJECTS_ERROR,
  PROJECTS_SAVED,
  PROJECTS_SAVING,
} from 'redux/actionTypes/projects';
import '__test__/test-utils/setupTests';

enableFetchMocks();
const mockStore = configureStore([thunk]);

describe('saveProject action', () => {
  const mockProject = {
    name: 'mockProject',
    description: 'Project',
    uuid: 'project-uuid',
    createdDate: '2021-01-01T00:00:00.000Z',
    lastModified: null,
    samples: [],
    lastAnalyzed: null,
  };

  const initialState = {
    projects: {
      ...initialProjectState,
      ids: [mockProject.uuid],
      meta: {
        activeprojectUuid: null,
      },
      [mockProject.uuid]: mockProject,
    },
  };

  beforeEach(() => {
    const response = new Response(JSON.stringify({ one: 'one' }));

    fetchMock.resetMocks();
    fetchMock.doMock();
    fetchMock.mockResolvedValueOnce(response);
  });

  it('Dispatches fetch correctly.', async () => {
    const store = mockStore(initialState);
    await store.dispatch(saveProject(mockProject.uuid, mockProject));

    expect(fetchMock).toHaveBeenCalledWith(
      `http://localhost:3000/v1/projects/${mockProject.uuid}`,
      {
        body: JSON.stringify(mockProject),
        headers: {
          'Content-Type': 'application/json',
        },
        method: 'PUT',
      },
    );
  });

  it('Dispatches a notification when fetch fails.', async () => {
    const errorMsg = 'some weird error that happened';

    fetchMock.resetMocks();
    fetchMock.mockResponse(JSON.stringify({ message: errorMsg }), { status: 400 });

    const store = mockStore(initialState);

    try {
      await store.dispatch(saveProject(mockProject.uuid, mockProject));
    } catch (e) {
      expect(e.statusCode).toEqual(400);
    }

    const actions = store.getActions();

    // First action sets up saving status
    expect(actions[0].type).toBe(PROJECTS_SAVING);

    // Second state saves error
    expect(actions[1].type).toBe(PROJECTS_ERROR);

    // Expect notification to be fired
    expect(pushNotificationMessage).toHaveBeenCalled();

    expect(actions).toMatchSnapshot();
  });

  it('Dispatches project pre and post actions correctly', async () => {
    const store = mockStore(initialState);
    await store.dispatch(saveProject(mockProject.uuid, mockProject));

    const actions = store.getActions();

    expect(actions.length).toEqual(2);
    expect(actions[0].type).toEqual(PROJECTS_SAVING);
    expect(actions[1].type).toEqual(PROJECTS_SAVED);

    expect(actions).toMatchSnapshot();
  });

  it('Does not dispatch pre and post actions if disabled', async () => {
    const store = mockStore(initialState);
    await store.dispatch(saveProject(mockProject.uuid, mockProject, false));

    const actions = store.getActions();
    expect(actions.length).toEqual(0);

    expect(actions).toMatchSnapshot();
  });
});
