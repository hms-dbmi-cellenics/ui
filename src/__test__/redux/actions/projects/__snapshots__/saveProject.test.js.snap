// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`saveProject action Dispatches a notification when fetch fails. 1`] = `
Array [
  Object {
    "payload": Object {
      "message": "Saving project...",
    },
    "type": "projects/saving",
  },
  Object {
    "payload": Object {
      "error": "some weird error that happened",
    },
    "type": "projects/error",
  },
]
`;

exports[`saveProject action Dispatches project pre and post actions correctly 1`] = `
Array [
  Object {
    "payload": Object {
      "message": "Saving project...",
    },
    "type": "projects/saving",
  },
  Object {
    "type": "projects/saved",
  },
]
`;

exports[`saveProject action Does not dispatch pre and post actions if disabled 1`] = `Array []`;
